galsim/phase_psf.py

```diff
@@ -0,0 +1,1017 @@
+                         sea-level.  [Default: 0.0]
+    @param time_step     Interval to use when advancing the screen in time in seconds.
+                         [Default: 0.03]
+    @param r0_500        Fried parameter setting the amplitude of turbulence; contributes to "size"
+                         of the resulting atmospheric PSF.  Specified at wavelength 500 nm, in units
+                         of meters.  [Default: 0.2]
+    @param L0_inv        Inverse outer scale in inverse meters, beyond which turbulence power
+                         spectrum smoothly approaches zero.  [Default: 1./25]
+    @param vx            x-component wind velocity in meters/second.  [Default: 0.]
+    @param vy            y-component wind velocity in meters/second.  [Default: 0.]
+    @param rng           Random number generator as a galsim.BaseDeviate().  If None, then use the
+                         clock time or system entropy to seed a new generator.  [Default: None]
+    """
+    def __init__(self, screen_size, screen_scale=None, altitude=0.0, time_step=0.03,
+                 r0_500=0.2, L0_inv=1./25.0, vx=0.0, vy=0.0, rng=None):
+
+        if screen_scale is None:
+            screen_scale = 0.5 * r0_500
+        super(AtmosphericScreen, self).__init__(screen_size, screen_scale, altitude)
+
+        if rng is None:
+            rng = galsim.BaseDeviate()
+        self.rng = rng
+        self.orig_rng = copy.deepcopy(rng)
+
+        self.time_step = time_step
+        self.altitude = altitude
+        self.r0_500 = r0_500
+        self.L0_inv = L0_inv
+        self.vx = vx
+        self.vy = vy
+
+    def advance_by(self, dt):
+        """Advance phase screen by specified amount of time.
+
+        @param dt  Amount of time in seconds by which to update the screen.
+        @returns   The actual amount of time updated, which will differ from `dt` when `dt` is not a
+                   multiple of self.time_step.
+        """
+        _nstep = int(np.ceil(dt/self.time_step))
+        for i in xrange(_nstep):
+            self.advance()
+        return _nstep*self.time_step  # return the time *actually* advanced
+
+    # Collect a few methods common to both FrozenAtmosphericScreen and ARAtmosphericScreen
+    def _freq(self):
+        """Return arrays of sampling frequencies for Discrete Fourier Transform of phase screen."""
+        fx = np.fft.fftfreq(self.npix, self.screen_scale)
+        return np.meshgrid(fx, fx)
+
+    def _init_screen(self, fx, fy):
+        """Initialize a phase screen from von Karman power spectrum."""
+        # Note the magic number 0.00058 is actually ... wait for it ...
+        # (5 * (24/5 * gamma(6/5))**(5/6) * gamma(11/6)) / (6 * pi**(8/3) * gamma(1/6)) / (2 pi)**2
+        # It nearly impossible to figure this out from a single source, but it can be derived from a
+        # combination of Roddier (1981), Sasiela (1994), and Noll (1976).  (These atmosphere people
+        # sure like to work alone... )
```

Comment:
Even with this entertaining comment, it would probably be better to give the magic number a name rather than use it bare in the code for self.psi.  Even if that name is just something like `magic_constant` or `vonKarman_psi_constant`.